### PR TITLE
Certora's 3rd Round of Formal Verification

### PR DESCRIPTION
**Pull request by teryanarmen**
Certora's 3rd round of verification for Open Zeppelin's contracts. In this round we verified:
Initializable.sol
GovernorPreventLateQuorum.sol
ERC1155Burnable.sol
ERC1155Pausable.sol
ERC1155Supply.sol

We also determined no formal verification needed for:
ERC1155Holder.sol
ERC1155Receiver.sol

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (Amxx) [inline .vscode/settings.json:1]: Is that file needed for running the certora workflow ? I'd guess not.

File: certora/helpers/DummyERC20B.sol [hunk 2]
```diff
@@ -0,0 +1,5 @@
+// SPDX-License-Identifier: agpl-3.0
+pragma solidity ^0.8.0;
+import "./DummyERC20Impl.sol";
+
+contract DummyERC20B is DummyERC20Impl {}
```
Comment (Amxx) [inline certora/helpers/DummyERC20B.sol:5]: Is that needed? is `DummyERC20A` not enough?

File: certora/helpers/DummyERC20Impl.sol [hunk 3]
```diff
@@ -0,0 +1,57 @@
+// SPDX-License-Identifier: agpl-3.0
+pragma solidity ^0.8.0;
+
+// with mint
+contract DummyERC20Impl {
```
Comment (Amxx) [inline certora/helpers/DummyERC20Impl.sol:5]: I'm missing the point of this contract. We have an ERC20 impl, why do we need that ?

File: certora/helpers/DummyERC20Impl.sol [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    uint256 t;
+    mapping (address => uint256) b;
+    mapping (address => mapping (address => uint256)) a;
+
+    string public name;
+    string public symbol;
+    uint public decimals;
```
Comment (Amxx) [inline certora/helpers/DummyERC20Impl.sol:12]: the ERC says decimals should be `uint8`, not `uint`

File: certora/scripts/Round1/verifyAll.sh [hunk 5]
```diff
@@ -16,7 +18,7 @@ do
                 certoraRun certora/harnesses/ERC20VotesHarness.sol certora/harnesses/$contractFile \
                 --link ${contractFile%.*}:token=ERC20VotesHarness \
                 --verify ${contractFile%.*}:certora/specs/$specFile "$@" \
-                --solc solc8.2 \
+                --solc solc \
```
Comment (Amxx) [inline certora/scripts/Round1/verifyAll.sh:21]: Other scripts, such as `certora/scripts/Round1/sanity.sh` and `certora/scripts/Round1/sanityTokens.sh` uses `solc8.2`.
We should unify that.

File: certora/harnesses/AccessControlHarness.sol [hunk 6]
```diff
@@ -0,0 +1,10 @@
+// SPDX-License-Identifier: MIT
+// OpenZeppelin Contracts (last updated v4.5.0) (access/AccessControl.sol)
+
+pragma solidity ^0.8.0;
+
+import "../munged/access/AccessControl.sol";
```
Comment (Amxx) [inline certora/harnesses/AccessControlHarness.sol:6]: munged contracts are not part of this PR. What are they? How are they generated? Can they be replaced with the `contracts-exposed` ?

File: .github/workflows/verify.yml [hunk 7]
```diff
@@ -0,0 +1,65 @@
+name: Certora
+
+on:
+  push:
+    branches:
+      - master
+      - main
+      - certora/erc20
+      - certora/erc1155ext
```
Comment (Amxx) [inline .github/workflows/verify.yml:9]: do we want to run that on every PR ?

File: .github/workflows/verify.yml [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+      - name: Install java
+        uses: actions/setup-java@v1
+        with: { java-version: "11", java-package: jre }
+
+      - name: Install certora
+        run: pip install certora-cli
+
+      - name: Install solc
+        run: |
+          wget https://github.com/ethereum/solidity/releases/download/v0.8.4/solc-static-linux
```
Comment (Amxx) [inline .github/workflows/verify.yml:31]: solidity version ?

File: .vscode/settings.json (hunk 1 above)
Comment (teryanarmen) [reply .vscode/settings.json:1]: It is not. I'll edit.

File: certora/helpers/DummyERC20B.sol (hunk 2 above)
Comment (teryanarmen) [reply certora/helpers/DummyERC20B.sol:5]: Sometimes we need to have 2 separate token contracts as they have their own state. I will check if its used and can remove it if its unnecessary.

File: certora/helpers/DummyERC20Impl.sol (hunk 3 above)
Comment (teryanarmen) [reply certora/helpers/DummyERC20Impl.sol:5]: This was created before as a basic implementation. We use this on most projects but are working on having a more diverse set up ERC20 implementations. i can replace it with OZ implementation if you'd like.

File: certora/helpers/DummyERC20Impl.sol (hunk 4 above)
Comment (teryanarmen) [reply certora/helpers/DummyERC20Impl.sol:12]: Will update or just use OZ implementation.

File: certora/scripts/Round1/verifyAll.sh (hunk 5 above)
Comment (teryanarmen) [reply certora/scripts/Round1/verifyAll.sh:21]: scripts ran by the ci downloads one version of solc and so uses `solc` executable. Some members on the team have a set of executables to make switching between versions easier. I can remove this script as its not ran by ci or leave it in case others from Certora, OZ, or the community try to add rules or run them.

File: .github/workflows/verify.yml (hunk 8 above)
Comment (teryanarmen) [reply .github/workflows/verify.yml:31]: version 0.8.4 is specified in the github link

File: .github/workflows/verify.yml (hunk 7 above)
Comment (teryanarmen) [reply .github/workflows/verify.yml:9]: Maybe master branch push and PRs? This way direct pushes to master are checked along with all PRs.

File: certora/harnesses/AccessControlHarness.sol (hunk 6 above)
Comment (teryanarmen) [reply certora/harnesses/AccessControlHarness.sol:6]: munged contracts are created on the fly with a using a `make` file. commands are explained in the file. Im not sure you mean by `contracts-exposed`. We can have a fixed folder of the changed contracts but that would make it difficult to update when changes are made to the original contracts.

File: .github/workflows/verify.yml (hunk 8 above)
Comment (Amxx) [reply .github/workflows/verify.yml:31]: Sorry my message was not clear (it was mostly an internal post-it of what we have to discuss). The real question in the line of:
- do we want to have a solc version hardcoded ? 
- was processes do we put in place to maintain that ?
- does this version have to match the scripts ?

**Comment by Amxx** [comment]
There are some merge conflict that I can't solve since I don't have write access to Certora:master
